### PR TITLE
Update shallow_copy_and_detach for nested tensor impls

### PR DESCRIPTION
**Pull request by drisspg**
# Summary
This change fixes a bug that was encountered when trying to add more backward formulas for nested tensor ops.  If a derivative is defined that stores the "result" for use in the backward the output of the forward op is saved using:
```
 if (grad_fn) {
    grad_fn->result_ = SavedVariable(result, true);
  }
```

SavedVariable calls a series of functions which in turn calls shallow_copy_and_detach and when https://github.com/pytorch/pytorch/blob/c179597753f20089056115542d3b9c74ab8b864c/c10/core/TensorImpl.cpp#L533 is hit this calls sizes_custom() which is not implemented and errors.  I also noticed that since the storage format is different for nested_tensor not `storage_ ` but instead two tensors that the we should actually be calling the NestedTensorImpl constructor.

This PR overrides shallow_copy_and_detach from the derived class and ensures that shallow copy works correctly.

## Update
- Added the softmax derivative in this PR because that is a direct use case that was blocked by not having shallow_copy_and_detach work correctly.

### CODE REVIEW

File: aten/src/ATen/NestedTensorImpl.cpp [hunk 1]
```diff
@@ -219,5 +238,52 @@ const char* NestedTensorImpl::tensorimpl_type_name() const {
...
+
+      copy_tensor_metadata(
+          /*src_impl=*/this,
+          /*dest_impl=*/impl.get(),
+          /*version_counter=*/std::forward<VariableVersion>(version_counter),
+          /*allow_tensor_metadata_change=*/allow_tensor_metadata_change);
+  return impl;
+}
+
+c10::intrusive_ptr<TensorImpl> NestedTensorImpl::shallow_copy_and_detach(
```
Comment (drisspg) [inline aten/src/ATen/NestedTensorImpl.cpp:274]: I don't think I need to specialize to the derived class of <NestedTensorImpl>

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -4575,6 +4572,8 @@
 
 - func: _softmax_backward_data(Tensor grad_output, Tensor output, int dim, ScalarType input_dtype) -> Tensor
   structured_delegate: _softmax_backward_data.out
+  dispatch:
+    NestedTensorCPU, NestedTensorCUDA: nested_softmax_backward
```
Comment (drisspg) [inline aten/src/ATen/native/native_functions.yaml:4576]: bringing this comment from Alban over here:
```
You need to leave the CompositeImplicitAutograd key here
```

Comment (drisspg) [reply aten/src/ATen/native/native_functions.yaml:4576]: Still can't get this to work if I add the CompositeImplicitAutograd key. I think because it is a structured delegate the dispatch table is in fact populated from the kernels defined from here:
```
- func: _softmax_backward_data.out(Tensor grad_output, Tensor output, int dim, ScalarType input_dtype, *, Tensor(a!) grad_input) -> Tensor(a!)
  structured: True
  dispatch:
    CPU: softmax_backward_cpu_out
    CUDA: softmax_backward_cuda_out
    MPS: softmax_backward_mps_out
```
so there is no `_softmax_backward_data` compositeImpllict kernel? or at least I am not sure what the name of that composite kernel is if it does exist, because it does not match the name `_softmax_backward_data`

Comment (drisspg) [reply aten/src/ATen/native/native_functions.yaml:4576]: I looked at the build directory and I do see `_softmax_backward_data` registered to the CPU, CUDA,
```
m.impl("_softmax_backward_data", TORCH_FN(wrapper__softmax_backward_data));
```
can't find a RegisterMPS.cpp function in build so I am just assuming that it is also getting registered there as well

Comment (drisspg) [reply aten/src/ATen/native/native_functions.yaml:4576]: @bdhirsh Does my above reasoning for why I get "undefined symbol" errors when trying to register composite implicit kernel for `_softmax_backward_data` fall in line with what you know about structured delegates?

Comment (zou3519) [reply aten/src/ATen/native/native_functions.yaml:4576]: > I feel like structured kernels should never be structured as that is obviously not composite compliant?

What exactly is the problem here? Is something bypassing dispatch?

Comment (drisspg) [reply aten/src/ATen/native/native_functions.yaml:4576]: Not to derail the convo but is the main issue here that the softmax_kernel for nested tensors was only registered/written for the functional version when in fact it should have been written for the out variant?

Comment (jbschlosser) [reply aten/src/ATen/native/native_functions.yaml:4576]: > Not to derail the convo but is the main issue here that the softmax_kernel for nested tensors was only registered/written for the functional version when in fact it should have been written for the out variant?

This seems right to me - I'd expect the functional version to be delegated to the out variant.

Comment (drisspg) [reply aten/src/ATen/native/native_functions.yaml:4576]: After the sage wisdom from Brian for native functions with a `structured_delegate` key. A `CompositeExplicitAutogradNonFunctional` kernel gets registered regardless of dispatch entries therefore it so not required to manual add here.

File: aten/src/ATen/NestedTensorImpl.cpp [hunk 3]
```diff
@@ -108,43 +108,62 @@ c10::DispatchKeySet generate_nested_key_set(at::Tensor buffer) {
...
-      "NestedTensorImpl buffer is required to be 1 dimensional but got a buffer with ",
-      buffer.dim(),
-      " dimensions.");
-  buffer_size_ = buffer_size_vec[0];
-
+      opt_sizes_(construct_opt_sizes(nested_size_tensor_)) {
   TORCH_WARN_ONCE(
       "The PyTorch API of nested tensors is in prototype stage and will change "
       "in the near future.");
-  TORCH_INTERNAL_ASSERT(buffer.is_cuda() || buffer.is_cpu(), "NestedTensorImpl buffer must be either CUDA or CPU but got ", buffer.device());
```
Comment (albanD) [inline aten/src/ATen/NestedTensorImpl.cpp:135]: You can still check the storage device if you want to preserve this.

File: aten/src/ATen/NestedTensorImpl.cpp [hunk 4]
```diff
@@ -219,5 +238,52 @@ const char* NestedTensorImpl::tensorimpl_type_name() const {
...
+      data_type_,
+      nested_size_tensor_,
+      nested_stride_tensor_,
+      offsets_);
+
+      copy_tensor_metadata(
+          /*src_impl=*/this,
+          /*dest_impl=*/impl.get(),
+          /*version_counter=*/std::forward<VariableVersion>(version_counter),
+          /*allow_tensor_metadata_change=*/allow_tensor_metadata_change);
```
Comment (albanD) [inline aten/src/ATen/NestedTensorImpl.cpp:270]: You need to manually copy the NestedTensorImpl specific fields here right?

File: test/test_nestedtensor.py [hunk 5]
```diff
@@ -482,6 +481,7 @@ def test_device_checks(self, device):
         self.assertEqual(nt.is_cuda, is_cuda)
 
     @dtypes(torch.float, torch.float16, torch.double)
+    @torch.inference_mode()
```
Comment (albanD) [inline test/test_nestedtensor.py:484]: Why?

File: aten/src/ATen/NestedTensorImpl.cpp (hunk 3 above)
Comment (drisspg) [reply aten/src/ATen/NestedTensorImpl.cpp:135]: Will add those checks

File: aten/src/ATen/NestedTensorImpl.cpp (hunk 4 above)
Comment (drisspg) [reply aten/src/ATen/NestedTensorImpl.cpp:270]: I created that new constructor which is used on line 257. Makes sure to copy over all nested specific meta data when creating the new tensor impl.  
`copy_tensor_metadata` will then copy over the version version_counter. 
And call:
https://github.com/pytorch/pytorch/blob/cd5efc6f082c81fd40712127638931b9e2e5ee69/c10/core/TensorImpl.cpp#L623

https://github.com/pytorch/pytorch/blob/cd5efc6f082c81fd40712127638931b9e2e5ee69/c10/core/TensorImpl.cpp#L596

There is some repeated work going on here. Since the above constructor populates all the nested specific fields. I think have this constructor is good since it is the most specific.

File: test/test_nestedtensor.py (hunk 5 above)
Comment (drisspg) [reply test/test_nestedtensor.py:484]: I think this was a rebase bug good catch. Will remove

**Review by albanD — APPROVED** [review]
SGTM !
